### PR TITLE
Vendor pluggy

### LINKED ISSUE #944: Vendoring pluggy?
In [this mail](https://mail.python.org/pipermail/pytest-dev/2015-August/003103.html), @nicoddemus said:

> I think Holger also said before that he would like to vendor pluggy into
pytest before releasing 2.8 thought. Is there anything here that we can do
to help?

I've thought about this, and I wonder what the rationale is - I actually see some reasons why it should *not* be vendored:

- Distributions generally dislike vendoring - see the appropriate documentation for [Fedora](https://fedoraproject.org/wiki/Packaging:No_Bundled_Libraries) or [Debian](http://www.debian.org/doc/debian-policy/ch-source.html#s-embeddedfiles) or [Gentoo](https://wiki.gentoo.org/wiki/Why_not_bundle_dependencies)
- It would help getting pluggy packaged for every distribution which packages pytest - which is a big plus for other projects which intend to use it in the future (like my own, qutebrowser)

cc @hpk42

### PR DESCRIPTION
**Pull request by nicoddemus**
Fix #944

### CODE REVIEW

**Comment by RonnyPfannschmidt** [comment]
thats the variant of vendoring i wanted to avoid

linux distributions will break this
also it basically copies the source tree over all the time

i would much prefer a variant of vendoring that would happen for wheel releases while keeping the sdist free of copies of source trees

i.e. a step in build_wheel that would remove the dependency and include the content

**Comment by The-Compiler** [comment]
How are Linux distributions more likely to break this, compared to (if I understand you right) not vendoring pluggy for sdists?

I trust them enough to be able to adjust the imports back to `import pluggy` - they'll introduce the issue of "the project under test uses a different version of pluggy than pytest" again, but that'd be the same (or even worse) than having that problem with every sdist, right?

I feel like doing vendoring in a different way than everyone else does it will just be more fragile :wink:

One possibility to make it easier for distributions (if we really want to do that) would be to have a `pluggy` shim which does something like:

```python
try:
    from _pytest.vendor.pluggy import *
except ImportError:
    from pluggy import *
```

then distributions would only have to delete the `vendor` directory.

Then again, if we explain the problem to them (pluggy might change in the future, projects under test might use pluggy as well, and being a test runner we're in a bit of a special situation), I wonder if distributions would be okay with pytest vendoring pluggy - even more so when seeing that the code came from pytest originally.

I'll mail `python-modules-team@lists.alioth.debian.org` (which is the maintainer group of the Debian package for pytest) to see what's their take on this is.

**Comment by RonnyPfannschmidt** [comment]
devian is a main offender of breaking vendor-ed things by de-vendoring,

which is why i'd like a wheel with the vendor-ed lib for pip usage, and explicit version pin in the sdists for the distributions

they are less likely to fuck up badly if the version pins are exact

**Comment by flub** [comment]
I think this is a simple way which, while not perfect, will work just fine.  Automatically doing it at sdist/bdsit_wheel time would be even nicer but is much more work.

In order to support de-vendoring for people I think the `setup.py install` command could be thought a `--no-vendor` option which would put pluggy back in install_requires and not install the vendored pluggy instead replacing it with a shim that uses the real pluggy `from pluggy import *` could work there.  That way people installing from the sdist can opt-out of the vendored pluggy.  That functionality would make this PR even better, but to me is not enough to hold it up.

I do think changing the vendoring behaviour between the wheel and sdist is a bad idea however, by default they should install the same thing.

**Comment by RonnyPfannschmidt** [comment]
good points

your proposal also is more refined

**Comment by hpk42** [comment]
I am fine with the approach -- what about @The-Compiler suggestion of "try-importing" say from a _pytest/_pluggy.py ? we could then simply import from _pytest._pluggy and it would either come from vendor, or if it doesn't exist, from the installed place.

**Comment by RonnyPfannschmidt** [comment]
i'd suggest to default to tracking it in a 
`pip install --no-compile --target _pytest/vendored_packages` folder and using a try import falling back to the global one

**Comment by nicoddemus** [comment]
@hpk42:
> we could then simply import from _pytest._pluggy and it would either come from vendor, or if it doesn't exist, from the installed place.

@RonnyPfannschmidt:
> i'd suggest to default to tracking it in a pip install --no-compile --target _pytest/vendored_packages folder and using a try import falling back to the global one

Done both. :smile:

**Comment by nicoddemus** [comment]
Rebased into `master` and squashed all commits into one.

File: _pytest/_pluggy.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+"""
+imports symbols from vendored "pluggy" if available, otherwise
+falls back to importing "pluggy" from the default namespace.
+"""
+
+try:
+    from _pytest.vendored_packages.pluggy import *  # noqa
```
Comment (hpk42) [inline _pytest/_pluggy.py]: why such a long name?  _pytest.vendor is what i would expect ...

**Comment by hpk42** [comment]
apart from the minor naming issue i'd be fine to merge!

**Comment by RonnyPfannschmidt** [comment]
we should add the vendoring folder to the coverage ignore list

**Comment by nicoddemus** [comment]
> we should add the vendoring folder to the coverage ignore list

Done

@hpk42:
> apart from the minor naming issue i'd be fine to merge!

It was named `vendor` originally, @RonnyPfannschmidt suggested to change it to `vendored_packages`, I guess to make it more obvious to distro maintainers.

**Comment by RonnyPfannschmidt** [comment]
i slightly prefer `vendored_packages` over `vendor` since it is closer to `site-packages`

**Comment by RonnyPfannschmidt** [comment]
a future folder rename is trivial and the change is good

@hpk42 we might want to update pluggy once more before 2.8 (for the version-conflict handling)

**Pull request merged**